nodejs 16.6 fails to build on darwin

trying to build nodejs 16.6 on darwin I get the following error

```
In file included from ../deps/v8/src/libplatform/tracing/trace-writer.cc:14:
../deps/v8/src/libplatform/tracing/recorder.h:17:10: fatal error: 'os/signpost.h' file not found
#include <os/signpost.h>
         ^~~~~~~~~~~~~~~
1 error generated.
make[1]: *** [tools/v8_gypfiles/v8_libplatform.target.mk:159: /private/tmp/nix-build-nodejs-16.6.0.drv-0/n
rm 2e0eb71b1e97d7e96d4a06b54146ae6d4d2eb9d0.intermediate
make: *** [Makefile:110: node] Error 2
```

This error seems to be fixed by adding `darwin.Libsystem` as a buildinput.

however trying to build after that I get another error

```
In file included from /nix/store/znv06z6pld0ivwddj52rishggwc2m94c-libcxx-7.1.0-dev/include/c++/v1/__hash_table:19:
/nix/store/znv06z6pld0ivwddj52rishggwc2m94c-libcxx-7.1.0-dev/include/c++/v1/cmath:321:9: error: no member named 'isless' in the global namespace
using ::isless;
      ~~^
/nix/store/znv06z6pld0ivwddj52rishggwc2m94c-libcxx-7.1.0-dev/include/c++/v1/cmath:322:9: error: no member named 'islessequal' in the global namespace
using ::islessequal;
      ~~^
/nix/store/znv06z6pld0ivwddj52rishggwc2m94c-libcxx-7.1.0-dev/include/c++/v1/cmath:323:9: error: no member named 'islessgreater' in the global namespace
using ::islessgreater;
      ~~^
/nix/store/znv06z6pld0ivwddj52rishggwc2m94c-libcxx-7.1.0-dev/include/c++/v1/cmath:324:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
/nix/store/znv06z6pld0ivwddj52rishggwc2m94c-libcxx-7.1.0-dev/include/c++/v1/cmath:325:9: error: no member named 'isunordered' in the global namespace
using ::isunordered;
      ~~^
13 errors generated.
```

which I think is coming from the warning
`WARNING: C++ compiler (CXX=clang++, 7.1.0) too old, need g++ 8.3.0 or clang++ 8.0.0`

darwin is still on llvm 7. There is an effort at the moment to update it to llvm 11, but it might take some more time.

I don't understand how to fix the c++ compiler version. Should it not be fixed to something higher than 7 rather than relying on the default stdenv one?

@gilligan @cko @marsam 

I'm happy to test whatever ideas you have

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/development/web/nodejs/v16.nix
+++ b/pkgs/development/web/nodejs/v16.nix
@@ -1,4 +1,4 @@
-{ callPackage, openssl, python3, enableNpm ? true }:
+{ callPackage, openssl, python3, fetchpatch, enableNpm ? true }:
 
 let
   buildNodejs = callPackage ./nodejs.nix {
@@ -10,4 +10,12 @@ in
     inherit enableNpm;
     version = "16.5.0";
     sha256 = "16dapj5pm2y1m3ldrjjlz8rq9axk85nn316iz02nk6qjs66y6drz";
+    patches = [
+      # Fix CVE-2021-22930 https://github.com/nodejs/node/pull/39423.
+      # It should be fixed by Node.js 16.6.0, but currently it fails to build on Darwin
+      (fetchpatch {
+        url = "https://github.com/nodejs/node/commit/9d950a0956bf2c3dd87bacb56807f37e16a91db4.patch";
+        sha256 = "1narhk5dqdkbndh9hg0dn5ghhgrd6gsamjqszpivmp33nl5hgsx3";
+      })
+    ];
   }

--- a/pkgs/development/web/nodejs/v16.nix
+++ b/pkgs/development/web/nodejs/v16.nix
@@ -8,6 +8,6 @@ let
 in
   buildNodejs {
     inherit enableNpm;
-    version = "16.6.0";
-    sha256 = "1ndrqx3k5m62r7nzl5za59m33bx10541n7xbaxxz7088ifh18msw";
+    version = "16.5.0";
+    sha256 = "16dapj5pm2y1m3ldrjjlz8rq9axk85nn316iz02nk6qjs66y6drz";
   }

